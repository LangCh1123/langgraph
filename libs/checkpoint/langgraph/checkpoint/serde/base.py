from typing import Any, Protocol


class SerializerProtocol(Protocol):
    """Protocol for serialization and deserialization of objects.

    - `dumps`: Serialize an object to bytes.
    - `loads`: Deserialize an object from bytes.

    Valid implementations include the `pickle`, `json` and `orjson` modules.
    """

    def dumps(self, obj: Any) -> tuple[str, bytes]:
        ...

    def loads(self, data: tuple[str, bytes]) -> Any:
        ...
