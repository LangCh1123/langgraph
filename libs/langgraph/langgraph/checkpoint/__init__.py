from langgraph_checkpoint.base import (
    BaseCheckpointSaver,
    Checkpoint,
    SerializerProtocol,
)
from langgraph_checkpoint.memory import MemorySaver

__all__ = [
    "BaseCheckpointSaver",
    "Checkpoint",
    "MemorySaver",
    "SerializerProtocol",
]
