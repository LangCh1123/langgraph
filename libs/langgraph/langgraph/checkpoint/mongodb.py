import pickle
from contextlib import AbstractContextManager, contextmanager
from types import TracebackType
from typing import Iterator, Optional

from langchain_core.runnables import RunnableConfig
from typing_extensions import Self
from langgraph.checkpoint.base import (
    BaseCheckpointSaver,
    Checkpoint,
    CheckpointAt,
    CheckpointTuple,
    SerializerProtocol,
)
from pymongo import MongoClient


class MongoDBSaver(BaseCheckpointSaver, AbstractContextManager):
    """A checkpoint saver that stores checkpoints in a MongoDB database.

    Note: This saver is not thread-safe. If you need to use it in a multi-threaded environment,
    you should create a separate saver instance for each thread. This is because the saver uses
    a single MongoDB connection, which is not thread-safe.

    Args:
        conn: The MongoDB connection object.
        serde: The serializer/deserializer to use for serializing and deserializing checkpoints.
        at: The checkpoint-at strategy to use for determining the checkpoint timestamp.

    Examples:

        >>> from pymongo import MongoClient
        >>> from langgraph.checkpoint.mongodb import MongoDBSaver
        >>> from langgraph.graph import StateGraph
        >>>
        >>> builder = StateGraph(int)
        >>> builder.add_node("add_one", lambda x: x + 1)
        >>> builder.set_entry_point("add_one")
        >>> builder.set_finish_point("add_one")
        >>> memory = MongoDBSaver.from_conn_string(MONGO_CONN_STRING)
        >>> graph = builder.compile(checkpointer=memory)
        >>> config = {"configurable": {"thread_id": "1"}}
        >>> graph.get_state(config)
        >>> result = graph.invoke(3, config)
        >>> graph.get_state(config)
        StateSnapshot(values=4, next=(), config={'configurable': {'thread_id': '1', 'thread_ts': '2024-05-04T06:32:42.235444+00:00'}}, parent_config=None)
    """

    serde = pickle
    conn: MongoClient
    is_setup: bool

    def __init__(
        self,
        conn: MongoClient,
        *,
        serde: Optional[SerializerProtocol] = None,
        at: Optional[CheckpointAt] = None,
    ) -> None:
        super().__init__(serde=serde, at=at)
        self.conn = conn
        self.is_setup = False

    @classmethod
    def from_conn_string(cls, conn_string: str) -> "MongoDBSaver":
        """Create a new MongoDBSaver instance from a MongoDB connection string."""
        return MongoDBSaver(conn=MongoClient(conn_string))

    def __enter__(self) -> Self:
        return self

    def __exit__(
        self,
        __exc_type: Optional[type[BaseException]],
        __exc_value: Optional[BaseException],
        __traceback: Optional[TracebackType],
    ) -> Optional[bool]:
        return self.conn.close()

    def setup(self) -> None:
        """Set up the MongoDB collection for storing checkpoints.
        This method creates an index on the `thread_id` and `thread_ts` fields to ensure that the
        collection is efficient for querying checkpoints by thread ID and timestamp.

        Note: This method is idempotent and can be called multiple times without any side effects.
        """
        if self.is_setup:
            return

        collection = self.conn["langgraph_checkpoint_db"][
            "langgraph_checkpoint_db_collection"
        ]
        collection.create_index([("thread_id", 1), ("thread_ts", 1)], unique=True)
        self.is_setup = True

    @contextmanager
    def cursor(self, transaction: bool = True):
        """Get a cursor to the MongoDB collection for storing checkpoints."""
        self.setup()
        collection = self.conn["langgraph_checkpoint_db"][
            "langgraph_checkpoint_db_collection"
        ]
        try:
            yield collection
        finally:
            if transaction:
                pass

    def get_tuple(self, config: RunnableConfig) -> Optional[CheckpointTuple]:
        """Get the latest checkpoint for the given thread ID.

        This method returns the latest checkpoint for the given thread ID. If the thread ID has
        multiple checkpoints, the latest one is returned. If the thread ID has no checkpoints, this
        method returns `None`.

        Args:
            config (RunnableConfig): The config to use for retrieving the checkpoint.

        Returns:
            Optional[CheckpointTuple]: The latest checkpoint for the given thread ID, or `None` if
            the thread ID has no checkpoints.

        """
        with self.cursor(transaction=False) as collection:
            if config["configurable"].get("thread_ts"):
                if value := collection.find_one(
                    {
                        "thread_id": config["configurable"]["thread_id"],
                        "thread_ts": config["configurable"]["thread_ts"],
                    }
                ):
                    return CheckpointTuple(
                        config,
                        self.serde.loads(value["checkpoint"]),
                        (
                            {
                                "configurable": {
                                    "thread_id": config["configurable"]["thread_id"],
                                    "thread_ts": value["parent_ts"],
                                }
                            }
                            if value.get("parent_ts")
                            else None
                        ),
                    )
            else:
                if value := collection.find_one(
                    {"thread_id": config["configurable"]["thread_id"]},
                    sort=[("thread_ts", -1)],
                ):
                    return CheckpointTuple(
                        {
                            "configurable": {
                                "thread_id": value["thread_id"],
                                "thread_ts": value["thread_ts"],
                            }
                        },
                        self.serde.loads(value["checkpoint"]),
                        (
                            {
                                "configurable": {
                                    "thread_id": value["thread_id"],
                                    "thread_ts": value["parent_ts"],
                                }
                            }
                            if value.get("parent_ts")
                            else None
                        ),
                    )

    def list(self, config: RunnableConfig) -> Iterator[CheckpointTuple]:
        """List checkpoints from the database.

        This method retrieves all checkpoints for the given thread ID in reverse chronological order.

        Args:
            config (RunnableConfig): The config to use for listing the checkpoints.

        Returns:
            Iterator[CheckpointTuple]: An iterator that yields all checkpoints for the given thread
            ID in reverse chronological order.
        """
        with self.cursor(transaction=False) as collection:
            for value in collection.find(
                {"thread_id": config["configurable"]["thread_id"]},
                sort=[("thread_ts", -1)],
            ):
                yield CheckpointTuple(
                    {
                        "configurable": {
                            "thread_id": value["thread_id"],
                            "thread_ts": value["thread_ts"],
                        }
                    },
                    self.serde.loads(value["checkpoint"]),
                    (
                        {
                            "configurable": {
                                "thread_id": value["thread_id"],
                                "thread_ts": value["parent_ts"],
                            }
                        }
                        if value.get("parent_ts")
                        else None
                    ),
                )

    def put(self, config: RunnableConfig, checkpoint: Checkpoint) -> RunnableConfig:
        """Save a checkpoint to the database.

        This method saves the given checkpoint to the database. The checkpoint is associated
        with the provided config and its parent config (if any).

        Args:
            config (RunnableConfig): The config to use for saving the checkpoint.
            checkpoint (Checkpoint): The checkpoint to save.

        Returns:
            RunnableConfig: The updated config with the thread ID and timestamp of the saved
            checkpoint.
        """
        with self.cursor() as collection:
            collection.insert_one(
                {
                    "thread_id": config["configurable"]["thread_id"],
                    "thread_ts": checkpoint["ts"],
                    "parent_ts": config["configurable"].get("thread_ts"),
                    "checkpoint": self.serde.dumps(checkpoint),
                }
            )
        return {
            "configurable": {
                "thread_id": config["configurable"]["thread_id"],
                "thread_ts": checkpoint["ts"],
            }
        }
