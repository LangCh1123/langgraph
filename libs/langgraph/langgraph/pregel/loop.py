import asyncio
from collections import deque
from contextlib import AsyncExitStack, ExitStack
from types import TracebackType
from typing import (
    TYPE_CHECKING,
    Any,
    AsyncContextManager,
    Callable,
    ContextManager,
    List,
    Literal,
    Mapping,
    Optional,
    Sequence,
    Tuple,
    Type,
    TypeVar,
    Union,
)

from langchain_core.callbacks import AsyncParentRunManager, ParentRunManager
from langchain_core.runnables import RunnableConfig
from typing_extensions import Self

from langgraph.channels.base import BaseChannel
from langgraph.channels.manager import (
    AsyncChannelsManager,
    ChannelsManager,
    create_checkpoint,
)
from langgraph.checkpoint.base import (
    BaseCheckpointSaver,
    Checkpoint,
    CheckpointMetadata,
    CheckpointTuple,
    PendingWrite,
    copy_checkpoint,
    empty_checkpoint,
)
from langgraph.constants import INPUT, INTERRUPT
from langgraph.managed.base import (
    AsyncManagedValuesManager,
    ManagedValueMapping,
    ManagedValuesManager,
)
from langgraph.pregel.algo import (
    PregelTaskWrites,
    apply_writes,
    increment,
    prepare_next_tasks,
    should_interrupt,
)
from langgraph.pregel.debug import map_debug_checkpoint, map_debug_tasks
from langgraph.pregel.executor import (
    AsyncBackgroundExecutor,
    BackgroundExecutor,
    Submit,
)
from langgraph.pregel.io import map_input, map_output_updates, map_output_values, single
from langgraph.pregel.types import PregelExecutableTask

if TYPE_CHECKING:
    from langgraph.pregel import Pregel


V = TypeVar("V")
INPUT_DONE = object()


class PregelLoop:
    input: Optional[Any]
    config: RunnableConfig
    checkpointer: Optional[BaseCheckpointSaver]
    checkpointer_get_next_version: Callable[[Optional[V]], V]
    checkpointer_put_writes: Optional[
        Callable[[RunnableConfig, Sequence[tuple[str, Any]], str], Any]
    ]
    checkpointer_put: Optional[
        Callable[[RunnableConfig, Checkpoint, CheckpointMetadata], Any]
    ]
    graph: "Pregel"

    submit: Submit
    channels: Mapping[str, BaseChannel]
    managed: ManagedValueMapping
    checkpoint: Checkpoint
    checkpoint_config: RunnableConfig
    checkpoint_metadata: CheckpointMetadata
    checkpoint_pending_writes: Optional[List[PendingWrite]]

    step: int
    status: Literal[
        "pending", "done", "interrupt_before", "interrupt_after", "out_of_steps"
    ]
    tasks: Sequence[PregelExecutableTask]
    stream: deque[Tuple[str, Any]]

    # public

    def mark_tasks_scheduled(self, tasks: Sequence[PregelExecutableTask]) -> None:
        """Mark tasks as scheduled, to be used by queue-based executors."""
        raise NotImplementedError

    def put_writes(self, task_id: str, writes: Sequence[tuple[str, Any]]) -> None:
        """Put writes for a task, to be read by the next tick."""
        self.checkpoint_pending_writes.extend((task_id, k, v) for k, v in writes)
        if self.checkpointer_put_writes is not None:
            self.submit(
                self.checkpointer_put_writes,
                {
                    **self.checkpoint_config,
                    "configurable": {
                        **self.checkpoint_config["configurable"],
                        "thread_ts": self.checkpoint["id"],
                    },
                },
                writes,
                task_id,
            )

    def tick(
        self,
        *,
        output_keys: Union[str, Sequence[str]] = None,
        interrupt_after: Optional[Sequence[str]] = None,
        interrupt_before: Optional[Sequence[str]] = None,
        manager: Union[None, AsyncParentRunManager, ParentRunManager] = None,
    ) -> bool:
        """Execute a single iteration of the Pregel loop.
        Returns True if more iterations are needed."""

        if self.status != "pending":
            raise RuntimeError("Cannot tick when status is no longer 'pending'")

        if self.input is not INPUT_DONE:
            self._first()
        elif all(task.writes for task in self.tasks):
            writes = [w for t in self.tasks for w in t.writes]
            # all tasks have finished
            apply_writes(
                self.checkpoint,
                self.channels,
                self.tasks,
                self.checkpointer_get_next_version,
            )
            # produce values output
            self.stream.extend(
                ("values", v)
                for v in map_output_values(output_keys, writes, self.channels)
            )
            # clear pending writes
            self.checkpoint_pending_writes.clear()
            # save checkpoint
            self._put_checkpoint(
                {
                    "source": "loop",
                    "writes": single(
                        map_output_updates(output_keys, self.tasks)
                        if self.graph.stream_mode == "updates"
                        else map_output_values(output_keys, writes, self.channels)
                    ),
                }
            )
            # after execution, check if we should interrupt
            if should_interrupt(self.checkpoint, interrupt_after, self.tasks):
                self.status = "interrupt_after"
                return False
        else:
            return False

        # check if iteration limit is reached
        if self.step > self.config["recursion_limit"]:
            self.status = "out_of_steps"
            return False

        # prepare next tasks
        self.tasks = prepare_next_tasks(
            self.checkpoint,
            self.graph.nodes,
            self.channels,
            self.managed,
            self.config,
            self.step,
            for_execution=True,
            manager=manager,
        )

        # if no more tasks, we're done
        if not self.tasks:
            self.status = "done"
            return False

        # if there are pending writes from a previous loop, apply them
        if self.checkpoint_pending_writes:
            for tid, k, v in self.checkpoint_pending_writes:
                if task := next((t for t in self.tasks if t.id == tid), None):
                    task.writes.append((k, v))

        # if all tasks have finished, re-tick
        if all(task.writes for task in self.tasks):
            return self.tick()

        # before execution, check if we should interrupt
        if should_interrupt(self.checkpoint, interrupt_before, self.tasks):
            self.status = "interrupt_before"
            return False

        # produce debug output
        self.stream.extend(("debug", v) for v in map_debug_tasks(self.step, self.tasks))

        return True

    # private

    def _first(self) -> None:
        # map inputs to channel updates
        if input_writes := deque(map_input(self.graph.input_channels, self.input)):
            # discard any unfinished tasks from previous checkpoint
            discard_tasks = prepare_next_tasks(
                self.checkpoint,
                self.graph.nodes,
                self.channels,
                self.managed,
                self.config,
                self.step,
                for_execution=True,
            )
            # apply input writes
            apply_writes(
                self.checkpoint,
                self.channels,
                discard_tasks + [PregelTaskWrites(INPUT, input_writes, [])],
                self.checkpointer_get_next_version,
            )
            # save input checkpoint
            self._put_checkpoint({"source": "input", "writes": self.input})
        else:
            # no input is taken as signal to proceed past previous interrupt
            self.checkpoint["versions_seen"].setdefault(INTERRUPT, {})
            for k in self.channels:
                if k in self.checkpoint["channel_versions"]:
                    version = self.checkpoint["channel_versions"][k]
                    self.checkpoint["versions_seen"][INTERRUPT][k] = version
        # done with input
        self.input = INPUT_DONE

    def _put_checkpoint(
        self,
        metadata: CheckpointMetadata,
    ) -> None:
        # assign step
        metadata["step"] = self.step
        # bail if no checkpointer
        if self.checkpointer_put is not None:
            # create new checkpoint
            self.checkpoint_metadata = metadata
            self.checkpoint = create_checkpoint(
                self.checkpoint, self.channels, self.step
            )
            # save it, without blocking
            self.submit(
                self.checkpointer_put,
                self.checkpoint_config,
                copy_checkpoint(self.checkpoint),
                self.checkpoint_metadata,
            )
            self.checkpoint_config = {
                **self.checkpoint_config,
                "configurable": {
                    **self.checkpoint_config["configurable"],
                    "thread_ts": self.checkpoint["id"],
                },
            }
            # produce debug output
            self.stream.extend(
                ("debug", v)
                for v in map_debug_checkpoint(
                    self.step,
                    self.checkpoint_config,
                    self.channels,
                    self.graph.stream_channels_asis,
                    self.checkpoint_metadata,
                )
            )
        # increment step
        self.step += 1


class SyncPregelLoop(PregelLoop, ContextManager):
    def __init__(
        self,
        input: Optional[Any],
        *,
        config: RunnableConfig,
        checkpointer: Optional[BaseCheckpointSaver],
        graph: "Pregel",
    ) -> None:
        self.stream = deque()
        self.stack = ExitStack()
        self.input = input
        self.config = config
        self.checkpointer = checkpointer
        self.checkpointer_get_next_version = (
            checkpointer.get_next_version if checkpointer else increment
        )
        self.checkpointer_put_writes = checkpointer.put_writes if checkpointer else None
        self.checkpointer_put = checkpointer.put if checkpointer else None
        self.graph = graph
        # TODO if managed values no longer needs graph we can replace with
        # managed_specs, channel_specs

    # context manager

    def __enter__(self) -> Self:
        saved = (
            self.checkpointer.get_tuple(self.config) if self.checkpointer else None
        ) or CheckpointTuple(self.config, empty_checkpoint(), {"step": -2}, None, [])
        self.checkpoint_config = {
            **self.config,
            **saved.config,
            "configurable": {
                **self.config.get("configurable", {}),
                **saved.config.get("configurable", {}),
            },
        }
        self.checkpoint = copy_checkpoint(saved.checkpoint)
        self.checkpoint_metadata = saved.metadata
        self.checkpoint_pending_writes = saved.pending_writes

        self.submit = self.stack.enter_context(BackgroundExecutor(self.config))
        self.channels = self.stack.enter_context(
            ChannelsManager(self.graph.channels, self.checkpoint, self.config)
        )
        self.managed = self.stack.enter_context(
            ManagedValuesManager(
                self.graph.managed_values_dict, self.config, self.graph
            )
        )
        self.status = "pending"
        self.step = self.checkpoint_metadata["step"] + 1

        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> Optional[bool]:
        del self.graph
        return self.stack.__exit__(exc_type, exc_value, traceback)


class AsyncPregelLoop(PregelLoop, AsyncContextManager):
    def __init__(
        self,
        input: Optional[Any],
        *,
        config: RunnableConfig,
        checkpointer: Optional[BaseCheckpointSaver],
        graph: "Pregel",
    ) -> None:
        self.stream = deque()
        self.stack = AsyncExitStack()
        self.input = input
        self.config = config
        self.checkpointer = checkpointer
        self.checkpointer_get_next_version = (
            checkpointer.get_next_version if checkpointer else increment
        )
        self.checkpointer_put_writes = (
            checkpointer.aput_writes if checkpointer else None
        )
        self.checkpointer_put = checkpointer.aput if checkpointer else None
        self.graph = graph
        # TODO if managed values no longer needs graph we can replace with
        # managed_specs, channel_specs

    # context manager

    async def __aenter__(self) -> Self:
        saved = (
            await self.checkpointer.aget_tuple(self.config)
            if self.checkpointer
            else None
        ) or CheckpointTuple(self.config, empty_checkpoint(), {"step": -2}, None, [])
        self.checkpoint_config = {
            **self.config,
            **saved.config,
            "configurable": {
                **self.config.get("configurable", {}),
                **saved.config.get("configurable", {}),
            },
        }
        self.checkpoint = copy_checkpoint(saved.checkpoint)
        self.checkpoint_metadata = saved.metadata
        self.checkpoint_pending_writes = saved.pending_writes

        self.submit = await self.stack.enter_async_context(AsyncBackgroundExecutor())
        self.channels = await self.stack.enter_async_context(
            AsyncChannelsManager(self.graph.channels, self.checkpoint, self.config)
        )
        self.managed = await self.stack.enter_async_context(
            AsyncManagedValuesManager(
                self.graph.managed_values_dict, self.config, self.graph
            )
        )
        self.status = "pending"
        self.step = self.checkpoint_metadata["step"] + 1

        return self

    async def __aexit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> Optional[bool]:
        del self.graph
        return await asyncio.shield(
            self.stack.__aexit__(exc_type, exc_value, traceback)
        )
