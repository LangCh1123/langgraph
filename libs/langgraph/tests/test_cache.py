import hashlib
import json
import operator
import time
from typing import (
    Annotated,
    Any,
    Literal,
    Optional,
    TypedDict,
    Union,
)

import pytest
from langchain_core.runnables import (
    RunnableConfig,
)
from syrupy import SnapshotAssertion

from langgraph.checkpoint.base import BaseCheckpointSaver
from langgraph.graph.state import StateGraph
from langgraph.pregel.types import CachePolicy


def custom_cache_key(input: Any, config: Optional[RunnableConfig] = None) -> str:
    """
    Generate a cache key based on the input and config.

    Args:
        input (Any): The input to the node.
        config (Optional[RunnableConfig]): The configuration for the node.

    Returns:
        str: A string key under which the output should be cached.
    """
    # Convert input to a JSON-serializable format
    if isinstance(input, dict):
        input_str = json.dumps(input, sort_keys=True)
    elif isinstance(input, (str, int, float, bool)):
        input_str = str(input)
    else:
        input_str = str(hash(input))

    # Extract relevant parts from the config
    config_str = ""
    if config:
        relevant_config = {
            "tags": config.get("tags", []),
            "metadata": config.get("metadata", {}),
        }
        config_str = json.dumps(relevant_config, sort_keys=True)
    # Combine input and config strings
    combined_str = f"{input_str}|{config_str}"
    # Generate a hash of the combined string
    return hashlib.md5(combined_str.encode("utf-8")).hexdigest()


@pytest.mark.parametrize("checkpointer_name", ["postgres"])
def test_in_one_fan_out_state_graph_waiting_edge_via_branch_with_cache(
    snapshot: SnapshotAssertion, request: pytest.FixtureRequest, checkpointer_name: str
) -> None:
    checkpointer: BaseCheckpointSaver = request.getfixturevalue(
        f"checkpointer_{checkpointer_name}"
    )

    def sorted_add(
        x: list[str], y: Union[list[str], list[tuple[str, str]]]
    ) -> list[str]:
        if isinstance(y[0], tuple):
            for rem, _ in y:
                x.remove(rem)
            y = [t[1] for t in y]
        return sorted(operator.add(x, y))

    class State(TypedDict, total=False):
        query: str
        answer: str
        docs: Annotated[list[str], sorted_add]

    workflow = StateGraph(State)
    call_count = 0

    def rewrite_query(data: State) -> State:
        return {"query": f'query: {data["query"]}'}

    def analyzer_one(data: State) -> State:
        return {"query": f'analyzed: {data["query"]}'}

    def retriever_one(data: State) -> State:
        nonlocal call_count
        call_count += 1
        return {"docs": ["doc1", "doc2"]}

    def retriever_two(data: State) -> State:
        time.sleep(0.1)
        return {"docs": ["doc3", "doc4"]}

    def qa(data: State) -> State:
        return {"answer": ",".join(data["docs"])}

    def rewrite_query_then(data: State) -> Literal["retriever_two"]:
        return "retriever_two"

    config = RunnableConfig(configurable={"thread_id": "1"})

    workflow.add_node("rewrite_query", rewrite_query)
    workflow.add_node("analyzer_one", analyzer_one)
    workflow.add_node(
        "retriever_one",
        retriever_one,
        cache=CachePolicy(custom_cache_key({"user_id": "a user"}, config)),
    )
    workflow.add_node("retriever_two", retriever_two)
    workflow.add_node("qa", qa)

    workflow.set_entry_point("rewrite_query")
    workflow.add_edge("rewrite_query", "analyzer_one")
    workflow.add_edge("analyzer_one", "retriever_one")
    workflow.add_conditional_edges("rewrite_query", rewrite_query_then)
    workflow.add_edge(["retriever_one", "retriever_two"], "qa")
    workflow.set_finish_point("qa")

    app_with_checkpointer = workflow.compile(
        checkpointer=checkpointer,
    )
    assert app_with_checkpointer.get_graph().draw_mermaid(with_styles=False) == snapshot

    interrupt_results = [
        c
        for c in app_with_checkpointer.stream(
            {"query": "what is weather in sf"}, config
        )
    ]
    assert interrupt_results == [
        {"rewrite_query": {"query": "query: what is weather in sf"}},
        {"analyzer_one": {"query": "analyzed: query: what is weather in sf"}},
        {"retriever_two": {"docs": ["doc3", "doc4"]}},
        {"retriever_one": {"docs": ["doc1", "doc2"]}},
        {"qa": {"answer": "doc1,doc2,doc3,doc4"}},
    ]
    # first time calling retriever_one()
    assert call_count == 1
    config = RunnableConfig(configurable={"thread_id": "2"})

    stream_results = [
        c
        for c in app_with_checkpointer.stream(
            {"query": "what is weather in sf"}, config
        )
    ]

    # retriever_one should be cached
    assert stream_results == [
        {"rewrite_query": {"query": "query: what is weather in sf"}},
        {"analyzer_one": {"query": "analyzed: query: what is weather in sf"}},
        {"retriever_two": {"docs": ["doc3", "doc4"]}},
        {"retriever_one": {"docs": ["doc1", "doc2"]}, "__metadata__": {"cached": True}},
        {"qa": {"answer": "doc1,doc2,doc3,doc4"}},
    ]

    # retriever_one call_count should not increase count because of cache
    assert call_count == 1

    # Cache is not used when checkpointer is not provided
    app_without_checkpointer = workflow.compile()
    assert (
        app_without_checkpointer.get_graph().draw_mermaid(with_styles=False) == snapshot
    )

    interrupt_results = [
        c
        for c in app_without_checkpointer.stream(
            {"query": "what is weather in sf"}, config
        )
    ]

    # retriever_one should not be cached
    assert interrupt_results == [
        {"rewrite_query": {"query": "query: what is weather in sf"}},
        {"analyzer_one": {"query": "analyzed: query: what is weather in sf"}},
        {"retriever_two": {"docs": ["doc3", "doc4"]}},
        {"retriever_one": {"docs": ["doc1", "doc2"]}},
        {"qa": {"answer": "doc1,doc2,doc3,doc4"}},  # This item is extra
    ]
    # retriever_one should not be cached, and call_count should increase
    assert call_count == 2

    # Test a new workflow with the same cache key
    new_workflow = StateGraph(State)
    config = RunnableConfig(configurable={"thread_id": "4"})
    new_workflow.add_node("rewrite_query", rewrite_query)
    new_workflow.add_node("analyzer_one", analyzer_one)
    new_workflow.add_node(
        "retriever_one",
        retriever_one,
        cache=CachePolicy(custom_cache_key({"user_id": "a user"}, config)),
    )
    new_workflow.add_node("retriever_two", retriever_two)
    new_workflow.add_node("qa", qa)

    new_workflow.set_entry_point("rewrite_query")
    new_workflow.add_edge("rewrite_query", "analyzer_one")
    new_workflow.add_edge("analyzer_one", "retriever_one")
    new_workflow.add_conditional_edges("rewrite_query", rewrite_query_then)
    new_workflow.add_edge(["retriever_one", "retriever_two"], "qa")
    new_workflow.set_finish_point("qa")

    app_with_checkpointer = new_workflow.compile(
        checkpointer=checkpointer,
    )
    assert app_with_checkpointer.get_graph().draw_mermaid(with_styles=False) == snapshot

    interrupt_results = [
        c
        for c in app_with_checkpointer.stream(
            {"query": "what is weather in sf"}, config
        )
    ]
    # retriever_one should be cached as long as the cache key is the same
    assert interrupt_results == [
        {"rewrite_query": {"query": "query: what is weather in sf"}},
        {"analyzer_one": {"query": "analyzed: query: what is weather in sf"}},
        {"retriever_two": {"docs": ["doc3", "doc4"]}},
        {"retriever_one": {"docs": ["doc1", "doc2"]}, "__metadata__": {"cached": True}},
        {"qa": {"answer": "doc1,doc2,doc3,doc4"}},
    ]
    # retriever_one call_count should not increase count because of cache
    assert call_count == 2

    # Test a new workflow with a different cache key
    another_new_workflow = StateGraph(State)
    config = RunnableConfig(configurable={"thread_id": "5"})
    another_new_workflow.add_node("rewrite_query", rewrite_query)
    another_new_workflow.add_node("analyzer_one", analyzer_one)
    another_new_workflow.add_node(
        "retriever_one",
        retriever_one,
        cache=CachePolicy(custom_cache_key({"user_id": "a different user"}, config)),
    )
    another_new_workflow.add_node("retriever_two", retriever_two)
    another_new_workflow.add_node("qa", qa)

    another_new_workflow.set_entry_point("rewrite_query")
    another_new_workflow.add_edge("rewrite_query", "analyzer_one")
    another_new_workflow.add_edge("analyzer_one", "retriever_one")
    another_new_workflow.add_conditional_edges("rewrite_query", rewrite_query_then)
    another_new_workflow.add_edge(["retriever_one", "retriever_two"], "qa")
    another_new_workflow.set_finish_point("qa")

    app_with_checkpointer = another_new_workflow.compile(
        checkpointer=checkpointer,
    )
    assert app_with_checkpointer.get_graph().draw_mermaid(with_styles=False) == snapshot

    interrupt_results = [
        c
        for c in app_with_checkpointer.stream(
            {"query": "what is weather in sf"}, config
        )
    ]
    # retriever_one should not be cached because the user_id is different
    assert interrupt_results == [
        {"rewrite_query": {"query": "query: what is weather in sf"}},
        {"analyzer_one": {"query": "analyzed: query: what is weather in sf"}},
        {"retriever_two": {"docs": ["doc3", "doc4"]}},
        {"retriever_one": {"docs": ["doc1", "doc2"]}},
        {"qa": {"answer": "doc1,doc2,doc3,doc4"}},
    ]
    # retriever_one call_count should increase count because of cache miss
    assert call_count == 3
