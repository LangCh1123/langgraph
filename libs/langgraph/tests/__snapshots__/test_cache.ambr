# serializer version: 1
# name: test_in_one_fan_out_state_graph_waiting_edge_via_branch_with_cache[postgres]
  '''
  graph TD;
  	__start__ --> rewrite_query;
  	analyzer_one --> retriever_one;
  	qa --> __end__;
  	retriever_one --> qa;
  	retriever_two --> qa;
  	rewrite_query --> analyzer_one;
  	rewrite_query -.-> retriever_two;
  
  '''
# ---
# name: test_in_one_fan_out_state_graph_waiting_edge_via_branch_with_cache[postgres].1
  '''
  graph TD;
  	__start__ --> rewrite_query;
  	analyzer_one --> retriever_one;
  	qa --> __end__;
  	retriever_one --> qa;
  	retriever_two --> qa;
  	rewrite_query --> analyzer_one;
  	rewrite_query -.-> retriever_two;
  
  '''
# ---
# name: test_in_one_fan_out_state_graph_waiting_edge_via_branch_with_cache[postgres].2
  '''
  graph TD;
  	__start__ --> rewrite_query;
  	analyzer_one --> retriever_one;
  	qa --> __end__;
  	retriever_one --> qa;
  	retriever_two --> qa;
  	rewrite_query --> analyzer_one;
  	rewrite_query -.-> retriever_two;
  
  '''
# ---
# name: test_in_one_fan_out_state_graph_waiting_edge_via_branch_with_cache[postgres].3
  '''
  graph TD;
  	__start__ --> rewrite_query;
  	analyzer_one --> retriever_one;
  	qa --> __end__;
  	retriever_one --> qa;
  	retriever_two --> qa;
  	rewrite_query --> analyzer_one;
  	rewrite_query -.-> retriever_two;
  
  '''
# ---
